src/OnChainSurplus.py

```diff
@@ -0,0 +1,113 @@
+gpv2Abi = json.loads(abiJson)
+#connecting to ethereum node
+w3 = Web3(Web3.HTTPProvider(f"https://mainnet.infura.io/v3/{INFURA_KEY}"))
+#creating an instance of GPv2 contract 
+contractInstance = w3.eth.contract(address=address, abi=gpv2Abi)
+
+class DecodedSettlement:
+    def __init__(self, tokens: List[str], clearing_prices: List[int], trades: List[Tuple[int, int, str, int, int, int, bytes, int, int, int, bytes]], interactions: List[List[Tuple[str, int, bytes]]]):
+        self.tokens = tokens
+        self.clearing_prices = clearing_prices
+        self.trades = trades
+        self.interactions = interactions
+
+    @classmethod
+    def new(cls, contractInstance, transaction):
+        # Decode the function input
+        decoded_input = contractInstance.decode_function_input(transaction.input)[1:]
+        # Convert the decoded input to the expected types
+        tokens = decoded_input[0]['tokens']
+        clearing_prices = decoded_input[0]['clearingPrices']
+        trades = decoded_input[0]['trades']
+        interactions = decoded_input[0]['interactions']
+
+        # Create and return a new instance of DecodedSettlement
+        return cls(tokens, clearing_prices, trades, interactions)
+
+
+def calculateSurplus(startBlock = None, endBlock = None, settlementHash = None):
```

Comment:
We should probably split `calculateSurplus` into separate functions:
1. One that returns the list of transaction hashes based on the args (and is totally decoupled from the logic contained in this file)
2. Second function that receives a list of hashes and calculates surplus for each of them.

Reply:
Agree with @sunce86's comment. And if we go that way, it would be good to rename the function and give a more accurate name.